Bhimes dev

TheByronHimes opened the pull request:
Next PR will be much much smaller, promise. 
This has all the basic functionality: post, patch, get, and (a few) tests to go along with it. The code isn't very polished, this was more about getting it working first. I'm also not confident in the file structure. I looked at some repos for reference, especially auth-service, but the way some functions are grouped in separate files, and where those files are placed, seemed arbitrary and therefore not significantly helpful at first.

Commit:
Adapted template placeholders and set up basic API

Commit:
Added simple tests

Commit:
Removed unreachable test code

Commit:
Added Enums to PcrTest and improved PcrTest tests

Commit:
Turn off "Render Whitespace" devcontainer config

Commit:
Remove strip_whitespace because OpenAPI complains

Commit:
NewSampleSubmission dto

Commit:
Move DAO logic to new file. Parameterize DAO tests

Commit:
Use Hexkit MongoDB

Remove DummyDao and set up actual MongoDB instance via Hexkit.

Remove DAO common Test class

Add error handling for update/retrieve API endpoints

Commit:
Updated devcontainer.json

Commit:
Address MyPy error report

Also put pre-commit-config back into static files list

KerstenBreuer committed:
fix

Commit:
Remove host address override in compose file

Commit:
Mongo Test Container

Implemented MongoDB test container in test_api.py

Removed MongoDummyDao

Added NewSampleSubmission to formalize post response.

Injected DAO dependency to path op functions.

Created fixture_client_with_db in test_api.py

Commit:
Removed dao.py

Moved DAO function to deps.py

Moved config for DAO to config.py

Made some tweaks to router.py with type-hinting & documentation

Commit:
Update openapi.yaml
Because I forgot to in the previous commit

Commit:
Add kafka and zookeeper to compose

Include akafka extras in hexkit install reqs

Commit:
Move models.py into core folder

Did this because I liked the cleaner structure of the dcs/ucs repos

KerstenBreuer commented on .pre-commit-config.yaml:
```diff
@@ -23,7 +23,7 @@ repos:
       - id: check-merge-conflict
       - id: check-symlinks
-      - id: check-executables-have-shebangs
+      #- id: check-executables-have-shebangs
```
Is this problem resolved? Can we reactivate the check?

KerstenBreuer replied:
seperate PR.

KerstenBreuer commented on covid_microservice/core/greeting.py:
```diff
@@ -18,7 +18,7 @@
+from .models import Greeting, GreetingExpression
 
 GREETINGS_EXPRESSIONS = [
     GreetingExpression(expression="Καλημέρα", language="Greek", isinformal=False),
```
Is this module still needed?

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+
+## *********************************************************
+## *********************************************************
+## *********************************************************
```
Why have you inserted these comments?

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+class Greeting(GreetingBase, MessageBase):
+    """A container storing a greeting for a specfic person incl. metadata"""
+
+    pass  # pylint: disable=unnecessary-pass
```
Is this still needed?

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+        """Returns a dictionary version of the object with only specified fields"""
+        if fields is None:
+            fields = list(self.__fields__.keys())
+        return {k: getattr(self, k) for k in fields}
```
BaseModels come with this function build in, see model.dict with the include argument: https://docs.pydantic.dev/usage/exporting_models/#modeldict

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+
+
+class Dto(BaseModel):
+    """Base Dto Class"""
```
We use the term DTO only in the context of a DAO. If you are looking for a more generic term, then you may use "dataclass" or the pydantic jargon "model".

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+        return {k: getattr(self, k) for k in fields}
+
+
+class StatusEnum(str, Enum):
```
Maybe beeing a bit more specific helps here. Moreover, Hungarian notation is usually not used in python because it conflicts with duck typing. So the type Enum shouldn't be part of the name.

```suggestion
class PcrTestStatus(str, Enum):
```

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+    FAILED = "failed"
+
+
+class TestResultEnum(str, Enum):
```
```suggestion
class PcrTestResult(str, Enum):
```

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+class TestResultEnum(str, Enum):
+    """Enumeration for PcrTest test_result values"""
+
+    UNSET = ""
```
```suggestion
    INCONCLUSIVE = "inconclusive"
```

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+    NEGATIVE = "negative"
+
+
+class NewSampleSubmission(BaseModel):
```
We usually call the creation models "SomethingCreation":
```suggestion
class SampleCreation(BaseModel):
```

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+
+
+class NewSampleResponse(BaseModel):
+    """Model to deliver access_token and sample_id after posting new sample"""
```
This already contains details on the implementation of the rest interface which should not be known in the core.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/api/main.py:
```diff
@@ -19,33 +19,18 @@
 @app.get("/", summary="Greet the world")
 async def index():
     """Greet the World"""
     return "Hello World."
```
Can be deleted.

KerstenBreuer commented on covid_microservice/api/deps.py:
```diff
@@ -0,0 +1,36 @@
+
+async def get_mongodb_pcrtest_dao():
+    """Return usable DAO"""
+    config = SamplesDaoFactoryConfig(
+        db_connection_str="mongodb://mongodb:27017", db_name="covid_db"
+    )
+    factory = MongoDbDaoFactory(config=config)
+    return await factory.get_dao(
+        name="samples", dto_model=PcrTest, id_field="access_token"
+    )
```
This is not part of the REST API and should live in a dedicated model or package.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/api/router.py:
```diff
@@ -0,0 +1,149 @@
+    response_model=dict,
+)
+async def get_sample(
+    access_token: str, mdao: MongoDbDaoNaturalId = Depends(get_mongodb_pcrtest_dao)
```
Instead of using the FastAPI dependency inject, consider using the one from hexkit that is based on the dependency-injector.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/api/router.py:
```diff
@@ -0,0 +1,149 @@
+        HTTPException: If the specified resource was not found.
+    """
+    try:
+        sample = await mdao.get_by_id(access_token.strip())
+        data_to_return = sample.dictify(
+            [
+                "patient_pseudonym",
+                "submitter_email",
+                "collection_date",
+                "status",
+                "test_result",
+                "test_date",
+            ]
+        )
+        return data_to_return
```
This is actually part of the core application logic and should not be in the API.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/api/router.py:
```diff
@@ -0,0 +1,149 @@
+    Returns:
+        NewSampleResponse: The access token and sample ID for the new sample.
+    """
+    sample_id = make_sample_id()
+    access_token = make_access_token()
+    pcrtest = PcrTest(
+        patient_pseudonym=data.patient_pseudonym,
+        submitter_email=data.submitter_email,
+        collection_date=data.collection_date,
+        sample_id=sample_id,
+        access_token=access_token,
+    )
+
+    await mdao.insert(pcrtest)
+    result = await mdao.get_by_id(access_token)
+    sample_response = NewSampleResponse(
+        access_token=result.access_token, sample_id=result.sample_id
+    )
```
Same as above. This is domain logic and should go to the core.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/api/router.py:
```diff
@@ -0,0 +1,149 @@
+        found, raises a 404 Not Found error.
+    """
+    try:
+        sample = await mdao.get_by_id(data.access_token)
+        sample.status = data.status
+        sample.test_result = data.test_result
+        sample.test_date = data.test_date
+        await mdao.update(sample)
+        return 204
```
Should go to the core.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/config.py:
```diff
@@ -17,16 +17,22 @@
+    collection_name = "samples"
+
+
+@config_from_yaml(prefix="covid_microservice")
```
Prefer something shorter here:

```suggestion
@config_from_yaml(prefix="cm")
```

KerstenBreuer commented on covid_microservice/config.py:
```diff
@@ -17,16 +17,22 @@
 
-    service_name: str = "my_microservice"  # Please adapt
+    service_name: str = "covid_microservice"
     language: SupportedLanguages = "Croatian"
```
This is not needed.

KerstenBreuer commented on covid_microservice/config.py:
```diff
@@ -17,16 +17,22 @@
+
+
+@config_from_yaml(prefix="covid_microservice")
 class Config(ApiConfigBase):
```
You need the Mongo configs from hexkit as well.

KerstenBreuer replied:
Separate PR.

KerstenBreuer commented on covid_microservice/core/samples.py:
```diff
@@ -12,10 +12,22 @@
-def test_dummy():
-    """Just makes the CI pass."""
-    assert True
+def help_update_sample(dto: PcrTest, dao: Any) -> None:
+    """Update sample"""
+    # result = dao.update(mapping = {"access_token": access_token}, new_data)
+    dao.update(dto=dto)
```
Why is this needed?

KerstenBreuer commented on tests/test_api.py:
```diff
@@ -0,0 +1,123 @@
+
+
+@pytest.mark.asyncio
+async def test_post_sample_nominal(client_with_db):
```
This is a bit nicer to get editor and linter support:

```suggestion
async def test_post_sample_nominal(client_with_db: TestClient):
```

KerstenBreuer commented on tests/test_api.py:
```diff
@@ -0,0 +1,123 @@
+@pytest.mark.asyncio
+async def test_post_sample_nominal(client_with_db):
+    """Test the /sample POST function"""
+    nss = {
```
What does nss stand for? Maybe consider longer descriptive name.

KerstenBreuer commented on tests/test_api.py:
```diff
@@ -0,0 +1,123 @@
+    assert isinstance(response, dict)
+    assert "sample_id" in response
+    assert "access_token" in response
+    assert len(response.keys()) == 2
```
In the tests, you do not check whether the correct result is achieved in the database.

KerstenBreuer commented on tests/test_api.py:
```diff
@@ -0,0 +1,123 @@
+        "collection_date": "2022-08-21T11:11",
+    }
+
+    with client_with_db:
```
Why do we need the with block here?

KerstenBreuer commented on covid_microservice/api/router.py:
```diff
@@ -0,0 +1,149 @@
+
+# GET /sample
+@sample_router.get(
+    "/sample/{access_token}",
```
The access token should not be used as sample ID but as a bearer token in the header.

KerstenBreuer replied:
Rest resources should be in plural.

KerstenBreuer commented on tests/test_api.py:
```diff
@@ -0,0 +1,123 @@
+    """Test what happens when a user searches with a bad access token"""
+
+    data = client_with_db.get(SAMPLE_URL + "eqfeqefew44")
+    assert data.json() == {"detail": "Not Found"}
```
Checking for status code 404 is probably enough here.

KerstenBreuer commented on covid_microservice/api/router.py:
```diff
@@ -0,0 +1,149 @@
+
+    await mdao.insert(pcrtest)
+    result = await mdao.get_by_id(access_token)
+    sample_response = NewSampleResponse(
```
In a POST you would typically expect either just the ID of the newly created resource or all details on that resource including the one you send in the request.

KerstenBreuer commented on tests/test_dto.py:
```diff
@@ -0,0 +1,250 @@
+        collection_date="2022-08-21T11:18",
+        access_token="1234567812345678",
+    )
+    assert issubclass(PcrTest, Dto)
+    assert isinstance(pcrtest, Dto)
```
Generally, we should try to check the behavior rather than implementation. Moreover, you don't have to test trivial things. So I think these tests are not required.

KerstenBreuer commented on tests/test_dto.py:
```diff
@@ -0,0 +1,250 @@
+    assert isinstance(pcrtest, PcrTest)
+
+
+@pytest.mark.xfail
+def test_pcrtest_with_too_long_name():
+    """test pcrtest patient pseudonym max length"""
+    PcrTest(
+        sample_id=make_sample_id(),
+        patient_pseudonym="sadfasdfasdfasdasdfkjalsdkjflaks\
+            djlfkasdlkfasldkfjalksdjfsdfdfdd",
+        submitter_email="test@test.com",
+        collection_date="2022-08-21T11:18",
+        access_token="2234567812345678",
+    )
```
Try to parameterize this and following tests.

KerstenBreuer commented on tests/test_dto.py:
```diff
@@ -0,0 +1,250 @@
+    )
+
+
+###########
+# New Sample Submission Tests
+###########
```
Avoid using comments to structure your modules.

KerstenBreuer reviewed (changes requested):
Really nice. What is currently missing is the hexagonal architecture.

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+    collection_date: str = Field(..., regex=r"^\d{4}(-\d{2}){2}T\d{2}:\d{2}[zZ]?$")
+
+
+class NewSampleResponse(BaseModel):
```
```suggestion
class Sample(SampleCreation):
```

KerstenBreuer commented on covid_microservice/core/models.py:
```diff
@@ -0,0 +1,131 @@
+class NewSampleResponse(BaseModel):
+    """Model to deliver access_token and sample_id after posting new sample"""
+
+    access_token: str
+    sample_id: int
```
```suggestion
    sample_id: int
    
    
class SampleAuth(Sample):
    access_token: str
```

Commit:
Separated API from domain logic

Applied THA design

Rewrote tests for models (API/happy path tests still missing)

Implemented access_token as bearer token in header

Renamed PcrTests to Samples
